### PR TITLE
Improve compiling time for the option `find_heuristic_step_size=True`

### PR DESCRIPTION
**Pull request by fehiepsi**
Addresses #585.

This optimization has been low-priority to me because I thought that `find_heuristic_step_size=True` is not important to use. But recently, @vanAmsterdam mentioned in #663 that our MCMC might have some regressions w.r.t. the last release. Though I suspected that random seed plays a role there, while waiting for the feedback in #663, it is still good to optimize compiling if users want to enable that option.

Tested on https://github.com/pyro-ppl/numpyro/issues/543, with `find_heuristic_step_size=True`, compiling time is reduced from almost 1 minute (master branch) to 12s (this PR). Note that only took about 1-2s with `find_heuristic_step_size=False`.

### CODE REVIEW

File: numpyro/infer/hmc_util.py [hunk 1]
```diff
@@ -355,9 +357,10 @@ def warmup_adapter(num_adapt_steps, find_reasonable_step_size=None,
     adaptation_schedule = jnp.array(build_adaptation_schedule(num_adapt_steps))
     num_windows = len(adaptation_schedule)
 
-    def init_fn(z, rng_key, step_size=1.0, inverse_mass_matrix=None, mass_matrix_size=None):
+    def init_fn(z_info, rng_key, step_size=1.0, inverse_mass_matrix=None, mass_matrix_size=None):
         """
-        :param z: Initial position of the integrator.
+        :param tuple z_info: A tuple containing the initial position of the integrator, its
```
Comment (neerajprad) [inline numpyro/infer/hmc_util.py:362]: nit: can we just reuse `IntegratorState` namedtuple for `z_info` and define `r` as None? Also, not important but we can also define a defaults argument in `NamedTuple` and set it to `None`. (See [workaround](https://stackoverflow.com/questions/11351032/namedtuple-and-default-values-for-optional-keyword-arguments) for python 3.6).

File: numpyro/infer/hmc_util.py [hunk 2]
```diff
@@ -390,25 +393,26 @@ def init_fn(z, rng_key, step_size=1.0, inverse_mass_matrix=None, mass_matrix_siz
...
         return HMCAdaptState(step_size, inverse_mass_matrix, mass_matrix_sqrt,
                              ss_state, mm_state, window_idx, rng_key)
 
-    def update_fn(t, accept_prob, z, state):
+    def update_fn(t, accept_prob, z_info, state):
         """
         :param int t: The current time step.
         :param float accept_prob: Acceptance probability of the current trajectory.
-        :param z: New position drawn at the end of the current trajectory.
+        :param z_info: A tuple contains new position drawn at the end of the current trajectory,
```
Comment (neerajprad) [inline numpyro/infer/hmc_util.py:414]: nit: can we reuse this z_info description in all the remaining docstrings too?

File: numpyro/infer/hmc_util.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/infer/hmc_util.py]: Thanks, that makes sense to me. I've made the change.

File: numpyro/infer/hmc_util.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/infer/hmc_util.py]: I have change `z_info` tuple to IntegratorState, so we can simply say "the new integrator state" here

File: test/test_hmc_util.py [hunk 3]
```diff
@@ -226,7 +226,8 @@ def potential_fn(q):
     fn = (jit(find_reasonable_step_size, static_argnums=(0, 1, 2))
           if jitted else find_reasonable_step_size)
     rng_key = random.PRNGKey(0)
-    step_size = fn(potential_fn, kinetic_fn, p_generator, init_step_size, m_inv, q, rng_key)
+    step_size = fn(potential_fn, kinetic_fn, p_generator, init_step_size, m_inv,
+                   (q, None, None), rng_key)
```
Comment (neerajprad) [inline test/test_hmc_util.py:230]: These tests also need to be modified, I think.

**Comment by vanAmsterdam** [comment]
> Addresses #585.
> 
> 
> 
> This optimization has been low-priority to me because I thought that `find_heuristic_step_size=True` is not important to use. But recently, @vanAmsterdam mentioned in #663 that our MCMC might have some regressions w.r.t. the last release. Though I suspected that random seed plays a role there, while waiting for the feedback in #663, it is still good to optimize compiling if users want to enable that option.
> 
> 
> 
> Tested on https://github.com/pyro-ppl/numpyro/issues/543, with `find_heuristic_step_size=True`, compiling time is reduced from almost 1 minute (master branch) to 12s (this PR). Note that only took about 1-2s with `find_heuristic_step_size=False`.

the regressions are probably due to a bug in my own code

File: test/test_hmc_util.py (hunk 3 above)
Comment (fehiepsi) [reply test/test_hmc_util.py]: Thanks Neeraj! I just fix them and will merge if tests pass due to a typo (missing a comma) at setup file, which removes the dependency of numpyro on `jax`.
